File: frameworks/motia/motia-py/playground/steps/todo/update_todo_step.py
"""Update Todo Step."""

from typing import Any

from motia import ApiRequest, ApiResponse, Stream, http, logger

todo_stream: Stream[dict[str, Any]] = Stream("todo")

config = {
    "name": "UpdateTodo",
    "description": "Update an existing todo item",
    "flows": ["todo-app"],
    "triggers": [
        http("PUT", "/todo"),
    ],
    "enqueues": [],
}


def handler(request: ApiRequest[dict[str, Any]]) -> ApiResponse[Any]:
    """Handle update todo request."""
    body = request.body or {}
    todo_id = body.get("todo_id")

    logger.info("Updating todo", body)

    if not todo_id:
        logger.error("todo_id is required")
        return ApiResponse(status=400, body={"error": "todo_id is required"})

    existing_todo = todo_stream.get("inbox", todo_id)

    if not existing_todo:
        logger.error("Todo not found")
        return ApiResponse(status=404, body={"error": "Todo not found"})

    updated_todo = {**existing_todo, **body}
    todo = todo_stream.set("inbox", todo_id, updated_todo)

    logger.info("Todo updated successfully", {"todo_id": todo_id})

    return ApiResponse(status=200, body=todo)

File: frameworks/motia/motia-js/playground/steps/todo/update-todo.step.ts
import { type Handlers, logger, type StepConfig, type UpdateOp } from 'motia'
import { z } from 'zod'
import { todoSchema } from './create-todo.step'
import { todoStream } from './todo.stream'

export const config = {
  name: 'UpdateTodo',
  description: 'Update an existing todo item',
  flows: ['todo-app'],
  triggers: [
    {
      type: 'http',
      method: 'PUT',
      path: '/todo/:todoId',
      bodySchema: z.object({
        description: z.string().optional(),
        dueDate: z.string().optional(),
        checked: z.boolean().optional(),
      }),
      responseSchema: {
        200: todoSchema,
        404: z.object({ error: z.string() }),
      },
    },
  ],
  enqueues: [],
  virtualSubscribes: ['todo-created'],
} as const satisfies StepConfig

export const handler: Handlers<typeof config> = async ({ request }) => {
  const { todoId } = request.pathParams || {}
  const body = request.body || {}

  logger.info('Updating todo', { todoId, body })

  const updateOps: UpdateOp[] = []

  if (body.checked !== undefined) {
    updateOps.push({ type: 'set', path: 'completedAt', value: body.checked ? new Date().toISOString() : undefined })
  }

  if (body.description !== undefined) {
    updateOps.push({ type: 'set', path: 'description', value: body.description })
  }

  if (body.dueDate !== undefined) {
    updateOps.push({ type: 'set', path: 'dueDate', value: body.dueDate })
  }

  if (updateOps.length === 0) {
    return { status: 400, body: { error: 'No fields to update' } }
  }

  const result = await todoStream.update('inbox', todoId, updateOps)

  if (!result.old_value) {
    return { status: 404, body: { error: `Todo with id ${todoId} not found` } }
  }

  logger.info('Todo updated successfully', { todoId })

  return { status: 200, body: result.new_value }
}
